db/range_tombstone_fragmenter.cc

```diff
@@ -0,0 +1,256 @@
+
+#include "util/kv_map.h"
+#include "util/vector_iterator.h"
+
+namespace rocksdb {
+
+namespace {
+
+struct ParsedInternalKeyComparator {
+  ParsedInternalKeyComparator(const InternalKeyComparator* c) : cmp(c) {}
+
+  bool operator()(const ParsedInternalKey& a,
+                  const ParsedInternalKey& b) const {
+    return cmp->Compare(a, b) < 0;
+  }
+
+  const InternalKeyComparator* cmp;
+};
+
+}  // anonymous namespace
+
+FragmentedRangeTombstoneIterator::FragmentedRangeTombstoneIterator(
+    std::unique_ptr<InternalIterator> unfragmented_tombstones,
+    const InternalKeyComparator& icmp, SequenceNumber snapshot)
+    : tombstone_cmp_(icmp.user_comparator()),
+      icmp_(&icmp),
+      ucmp_(icmp.user_comparator()) {
+  if (unfragmented_tombstones == nullptr) {
+    pos_ = tombstones_.end();
+    return;
+  }
+  bool is_sorted = true;
+  int num_tombstones = 0;
+  InternalKey pinned_last_start_key;
+  Slice last_start_key;
+  for (unfragmented_tombstones->SeekToFirst(); unfragmented_tombstones->Valid();
+       unfragmented_tombstones->Next(), num_tombstones++) {
+    if (num_tombstones > 0 &&
+        icmp_->Compare(last_start_key, unfragmented_tombstones->key()) > 0) {
+      is_sorted = false;
+      break;
+    }
+    if (unfragmented_tombstones->IsKeyPinned()) {
+      last_start_key = unfragmented_tombstones->key();
+    } else {
+      pinned_last_start_key.DecodeFrom(unfragmented_tombstones->key());
+      last_start_key = pinned_last_start_key.Encode();
+    }
+  }
+  if (is_sorted) {
+    FragmentTombstones(std::move(unfragmented_tombstones), snapshot);
+    return;
+  }
+
+  std::vector<std::string> keys, values;
+  keys.reserve(num_tombstones);
+  values.reserve(num_tombstones);
+  for (unfragmented_tombstones->SeekToFirst(); unfragmented_tombstones->Valid();
+       unfragmented_tombstones->Next()) {
+    keys.emplace_back(unfragmented_tombstones->key().data(),
+                      unfragmented_tombstones->key().size());
+    values.emplace_back(unfragmented_tombstones->value().data(),
+                        unfragmented_tombstones->value().size());
+  }
+  auto iter = std::unique_ptr<VectorIterator>(
+      new VectorIterator(std::move(keys), std::move(values), icmp_));
+  FragmentTombstones(std::move(iter), snapshot);
+}
+
+void FragmentedRangeTombstoneIterator::FragmentTombstones(
+    std::unique_ptr<InternalIterator> unfragmented_tombstones,
+    SequenceNumber snapshot) {
+  Slice cur_start_key(nullptr, 0);
+  auto cmp = ParsedInternalKeyComparator(icmp_);
+
+  // Stores the end keys and sequence numbers of range tombstones with a
+  // start key of cur_start_key. Provides an ordering by end key for use
+  // in flush_current_tombstones.
+  std::set<ParsedInternalKey, ParsedInternalKeyComparator> cur_end_keys(cmp);
+
+  // Given the next start key in unfragmented_tombstones,
+  // flush_current_tombstones writes every tombstone fragment that starts
+  // and ends with a key before next_start_key.
```

Comment:
is "starts at or after `cur_start_key`" another constraint on the fragments output by this function?

Reply:
Yeah, it is another constraint (`cur_start_key` is updated in `flush_current_tombstones`, but based on its value when the function is called, that constraint is true).